base_control/Movement.ino

```diff
@@ -0,0 +1,162 @@
+Movement::Movement() {
+                    //id,d1,d2,p1,e1,e2
+  this->B_left  = Motor(1,51,50,4,3,15);
+  this->F_left  = Motor(2,37,36,5,2,22);
+  this->B_right = Motor(3,53,52,6,19,14);
+  this->F_right = Motor(4,25,24,7,18,23);
+}
+
+void Movement::pwm(int pwm) {
+  this->B_right.changePWM(pwm);
+  this->F_right.changePWM(pwm);
+  this->B_left.changePWM(pwm);
+  this->F_left.changePWM(pwm);
+}
+
+//PID
+void  Movement::constantSpeed(){
+  this->F_right.constantSpeed();
+  this->F_left.constantSpeed();
+  this->B_left.constantSpeed();
+  this->B_right.constantSpeed();
+}
+
+double Movement::getVelocity(Motor A){
+  double revolutions=A.ticks/PULSES_PER_REVOLUTION;
+  return (revolutions*WHEEL_DIAMETER*PI_C)/(TIME_VELOCITY_SAMPLE/1000);
+}
+void Movement::calcVelocityaux(Motor &A){
+  if(millis()-A.VelocityTiming < TIME_VELOCITY_SAMPLE)
+    return;
+  
+  A.velocity=getVelocity(A);
+  A.lastticks=A.ticks;  
+  A.ticks=0;
+  A.VelocityTiming=millis();
+
+}
+void Movement::calcVelocity(){
+  calcVelocityaux(this->B_right);
+  calcVelocityaux(this->F_right);
+  calcVelocityaux(this->B_left);
+  calcVelocityaux(this->F_left);
+}
+
+double Movement::getTargetTicks(){
+  double ticks = getTargetVelocity();
+  ticks=ticks * (TIME_VELOCITY_SAMPLE/1000);
+  ticks=ticks/(WHEEL_DIAMETER*PI_C);  
+  return ceil(ticks*PULSES_PER_REVOLUTION);
+}
+
+double Movement::getTargetVelocity(){
+  return sqrt(this->dX*this->dX+this->dY*this->dY);
+}
+double Movement::getTargetAngle(){
+  return atan(this->dY/this->dX);
+}
```

Comment:
if dX and dY never change, why keep doing all the math? Or what am I missing?

Reply:
in the new commit this variables are changing, I was missing this implementation. Review cmdVelocity method in Odometry class